packages/datastore-lock/src/datastore-lock.ts

```diff
@@ -0,0 +1,171 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+import {v4 as uuidv4} from 'uuid';
+import {Datastore, Key} from '@google-cloud/datastore';
+import {logger} from 'gcf-utils';
+
+const DEFAULT_LOCK_EXPIRY = 20 * 1000; // 20 seconds
+const MAX_LOCK_EXPIRY = 60 * 1000; // 60 seconds
+const DEFAULT_LOCK_ACQUIRE_TIMEOUT = 120 * 1000; // 120 seconds
+const BACKOFF_INITIAL_DELAY = 2 * 1000; // 1 seconds
+const BACKOFF_MAX_DELAY = 10 * 1000; // 10 seconds
+const DATASTORE_LOCK_ERROR_NAME = 'DatastoreLockError';
+
+enum AcquireResult {
+  Success = 1,
+  Failure,
+  LockActive,
+}
+
+const sleep = (ms: number) => {
+  return new Promise(r => setTimeout(r, ms));
+};
+
+/**
+ * A simple lock backed by Cloud Datastore
+ */
+export class DatastoreLock {
+  private uniqueId: string;
+  private kind: string;
+  private target: string;
+  private datastore: Datastore;
+  private key: Key;
+  private lockExpiry: number;
+  private lockAcquireTimeout: number;
+
+  constructor(
+    lockId: string,
+    target: string,
+    lockExpiry: number = DEFAULT_LOCK_EXPIRY,
+    lockAcquireTimeout: number = DEFAULT_LOCK_ACQUIRE_TIMEOUT
+  ) {
+    if (lockExpiry > MAX_LOCK_EXPIRY) {
+      throw new Error(
+        `lockExpiry is too long, max is ${MAX_LOCK_EXPIRY}, ` +
+          `given ${lockExpiry}`
+      );
+    }
+    this.datastore = new Datastore();
+    this.kind = `ds-lock-${lockId}`;
+    this.target = target;
+    this.key = this.datastore.key([this.kind, this.target]);
+    this.uniqueId = uuidv4();
+    this.lockExpiry = lockExpiry;
+    this.lockAcquireTimeout = lockAcquireTimeout;
+  }
+
+  /**
+   * Acquire the lock.
+   */
+  public async acquire(): Promise<boolean> {
+    const startTime = Date.now();
+    // Avoid multiple clients waiting for the same period of time.
+    let waitTime = BACKOFF_INITIAL_DELAY + Math.floor(Math.random() * 1000);
+    const maxDelay = BACKOFF_MAX_DELAY + Math.floor(Math.random() * 1000);
+
+    while (Date.now() - startTime < this.lockAcquireTimeout) {
+      const result = await this._acquire();
+      if (result === AcquireResult.Success) {
+        return true;
+      }
+      if (result === AcquireResult.Failure) {
+        logger.debug(
+          `Failed to acquire the lock, retrying after ${waitTime} miliseconds`
+        );
+      }
+      if (result === AcquireResult.LockActive) {
+        // The lock is active, wait and retry.
+        logger.debug(
+          `The lock is active, retrying after ${waitTime} miliseconds`
+        );
+      }
+      await sleep(waitTime);
+      waitTime = Math.min(
+        maxDelay,
+        waitTime + Math.floor(Math.random() * 2000)
+      );
+    }
+    // Timeout
+    return false;
+  }
+
+  private async _acquire(): Promise<AcquireResult> {
```

Comment:
Fun fact, I learned recently that `private` fields aren't actually private outside of the context of TypeScript, so when I actually wanted private methods, I ended up using symbols as method names:

https://github.com/yargs/yargs/blob/master/lib/yargs-factory.ts#L88

I'm not suggesting we change anything, just wanted to share this surprising TypeScript behavior.

Reply:
Ah it makes sense.